src/main/java/io/github/thebusybiscuit/cscorelib2/blocks/BlockPosition.java

```diff
@@ -0,0 +1,179 @@
+package io.github.thebusybiscuit.cscorelib2.blocks;
+
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.Chunk;
+import org.bukkit.block.Block;
+
+import java.lang.ref.WeakReference;
+import java.util.Objects;
+
+/**
+ * This is the position of a block in a World. Using this class as opposed to {@link Location} is much better because it
+ * has a lower memory footprint and doesn't contain useless data such as yaw and pitch.<br />
+ * This is 12 bytes in memory whereas {@link Location} is 36 bytes.
+ * (Note, this is not accounting for things like object headers)
+ */
+public final class BlockPosition {
+
+    private final WeakReference<World> world;
+    private final long position;
+
+    public BlockPosition(World world, long position) {
+        Objects.requireNonNull(world, "World cannot be null!");
+
+        this.world = new WeakReference<>(world);
+        this.position = position;
+    }
+
+    public BlockPosition(World world, int x, int y, int z) {
+        Objects.requireNonNull(world, "World cannot be null!");
+
+        this.world = new WeakReference<>(world);
+        this.position = ((long) (x & 0x3FFFFFF) << 38) | ((long) (z & 0x3FFFFFF) << 12) | (long) (y & 0xFFF);
+    }
```

Comment:
We got Lombok in this project, let's rather use the annotation from there for this.

Reply:
For the constructor? Lombok can't handle the constructor like it's required here.

Reply:
Why not? @NonNull does pretty much the exact same and we use it in every other class too...

Reply:
Oh you mean in replacement of the requireNotNull? Can do

Reply:
Yes, I thought I only selected that line... whoops